### DISCUSSION TITLE
Exposed gameserver ports

### DISCUSSION THREAD

**Discussion by svetch** [jeremylvln/Shulker#329] [Q&A] [answered]
Hello, we have begun the process of migrating our legacy servers to Kubernetes using Shulker. However, we've noticed that the game server ports are being exposed to the host. Could you explain why this is happening and provide guidance on how to prevent the game server ports from being exposed?

**Comment by jeremylvln** [comment]
Hi!

Really good question. It is actually an Agones behavior ([see here](https://agones.dev/site/docs/faq/#how-is-traffic-routed-from-the-allocated-port-to-the-gameserver-container)).

It's a bit unusual I agree with you. It is mainly because in most games, users connects directly to the servers and are not routed through a proxy. So having a dedicated port on the host avoids networks hops and so latency and overload on Kubernetes. Here we are not interested in this fast path, but I don't think we have control on that (but if we do I really should disable that!).

One way of "solving" this is to have a "private" Kubernetes Cluster by having your game server nodes not exposed through the Internet directly (or with firewall rules). But I can understand that all clusters cannot be private especially if you are provisioning your own without cloud providers :/

**Reply by jeremylvln** [reply]
My previous message was me answering on the phone because it was really late in Paris. After looking deeper into this, it seems that there is a way of preventing this. It's not very explicit in the documentation, but it's worth trying. I'll try on my end and get back to you with the results :)

**Reply by jeremylvln** [reply]
It worked well, I just released a fix version (v0.5.1), it should not happen anymore (and I'm so happy for that)

**Reply by svetch** [reply]
Thank you very much for your availability and quick response!

**Reply by jeremylvln** [reply]
You're welcome, do not hesitate to get back to me if something is wrong for you. I would happily help you migrating :)

**Reply by svetch** [reply]
We have problem with your fix, after we upgrade the shulker to v0.5.1 we find out an error on proxy server:

```
[17:35:32 ERROR] [io.fabric8.kubernetes.client.informers.impl.cache.SharedProcessor]: agones.dev/v1/namespaces/shulker-system/gameservers failed invoking io.shulkermc.proxyagent.adapters.kubernetes.ImplKubernetesGatewayAdapter$createEventHandler$1@2200fe48 event handler: Index 0 out of bounds for length 0
java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
at jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64) ~[?:?]
at jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70) ~[?:?]
at jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266) ~[?:?]
at java.util.Objects.checkIndex(Objects.java:359) ~[?:?]
at java.util.ArrayList.get(ArrayList.java:427) ~[?:?]
at io.shulkermc.proxyagent.services.ServerDirectoryService.registerServer(ServerDirectoryService.kt:43) ~[?:?]
at io.shulkermc.proxyagent.services.ServerDirectoryService.access$registerServer(ServerDirectoryService.kt:8) ~[?:?]
at io.shulkermc.proxyagent.services.ServerDirectoryService$1.invoke(ServerDirectoryService.kt:18) ~[?:?]
at io.shulkermc.proxyagent.services.ServerDirectoryService$1.invoke(ServerDirectoryService.kt:15) ~[?:?]
at io.shulkermc.proxyagent.adapters.kubernetes.ImplKubernetesGatewayAdapter$createEventHandler$1.onUpdate(ImplKubernetesGatewayAdapter.kt:89) ~[?:?]
at io.fabric8.kubernetes.client.informers.impl.cache.ProcessorListener$UpdateNotification.handle(ProcessorListener.java:92) ~[?:?]
at io.fabric8.kubernetes.client.informers.impl.cache.ProcessorListener.add(ProcessorListener.java:50) ~[?:?]
at io.fabric8.kubernetes.client.informers.impl.cache.SharedProcessor.lambda$distribute$0(SharedProcessor.java:91) ~[?:?]
at io.fabric8.kubernetes.client.informers.impl.cache.SharedProcessor.lambda$distribute$1(SharedProcessor.java:114) ~[?:?]
at io.fabric8.kubernetes.client.utils.internal.SerialExecutor.lambda$execute$0(SerialExecutor.java:58) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
at java.lang.Thread.run(Thread.java:833) [?:?]
```

We figured out that the ServerDirectoryService at 43 line try to access a port from list that is empty.

Our status object on gameserver crd:

```
Status:                                                                                                                                                                                                       
   Address:  NODE_IP
   Addresses:                                                                                                                                                                                                  
     Address:  NODE_IP                                                                                                                                                                                  
     Type:     InternalIP                                                                                                                                                                                      
     Address:  NODE_HOSTNAME                                                                                                                                                                                           
     Type:     Hostname                                                                                                                                                                                        
   Eviction:                                                                                                                                                                                                   
     Safe:              OnUpgrade                                                                                                                                                                              
   Immutable Replicas:  1                                                                                                                                                                                      
   Node Name:           NODE_HOSTNAME                                                                                                                                                                                  
   Players:             <nil>                                                                                                                                                                                  
   Ports:                                                                                                                                                                                                      
   Reserved Until:  <nil>                                                                                                                                                                                      
   State:           Ready
```

What we expect that the address is not the node address but the pod address and hostname. Could you please evaluate if this is possible, and if not how could we work around this behaviour?

**Reply by jeremylvln** [reply]
I must have seen this coming, very sorry about that. I went too fast. I see what the issue is, I will work on that today!

**Reply by jeremylvln** [reply]
So, I'm afraid I don't have a solution for now. I will disable the latest release as it's clearly broken. I've opened an issue on Agones ([see here](https://github.com/googleforgames/agones/issues/3590)) to ask for the feature of not binding a node port per game server.

I see in your snippet that Agones used your node internal IP. That's good because it means that your players cannot reach your servers directly. But I agree that this is not ideal.

Let's wait for the Agones team to respond to my issue, they are usually quite fast to react.

**Reply by svetch** [reply]
Our scenario is somewhat complex, as the node's IP in our case is not internal; rather, it is the node's public IP. This is because the worker nodes (on bare metal) are not on the same physical network as the control plane (which is on the cloud).

**Reply by jeremylvln** [reply]
Oh I see, this gives way more sense, thank you.

I've reverted my fix (that's why there was a quick release some hours ago) for to unlock most of the users (including me lol). But it can't work for you, there is no way you want the ports to be bound to the node public IP. There is an alternative that was suggested by an Agones contributor, but it needs a bit of work on my end and it was quite a busy week for me. I'll try to work on that this week-end.

**Reply by svetch** [reply]
Our plan is to complete the migration of our Velocity proxy and the lobbies by the end of this month. Only after we start migrating another around 10 types of servers. So we have time for this fix. Thank you for your support!

I've discovered another bug related to this: when attempting to run multiple lobbies, such as setting replicas to 3, only one lobby starts due to this error (from argocd pipeline):
```
one or more objects failed to apply, reason: Operation cannot be fulfilled on minecraftserverfleets.shulkermc.io "lobby": the object has 
been modified; please apply your changes to the latest version and try again,admission webhook "validations.agones.dev" denied the 
request: GameServer.agones.dev "lobby-p4l2h-fw499" is invalid: spec.ports[0].hostPort: Forbidden: HostPort cannot be specified with a 
Dynamic or Passthrough PortPolicy (retried 5 times).
```

I hope autoscaling functions properly; I haven't tested it yet, only attempted to increase the replica count of MinecraftServerFleet.

**Reply by jeremylvln** [reply, marked as answer]
I have done an implementation that should fix the issue. It's merged, and I started a release. Let me know if everything is okay for you and your use cases.

Now, no ports will be bound, and Shulker will use headless internal DNS to route the proxies to the servers.

I'm not a big fan of this solution as it will load the DNS servers, but it's acceptable until I find something better.

**Reply by svetch** [reply]
Thank you for your work and your availability!
